gpu skinning

CrDem opened the pull request.

CrDem committed:
glTF skinning + bone animations

CrDem committed:
get back

CrDem committed:
anim supply

CrDem committed:
origin pulled

CrDem committed:
cuda test

CrDem committed:
skinning on gpu

cuda kernel for skinning in OptixRender

CrDem committed:
jointMat alloc optimized + requested changes

CrDem committed:
logger link fixed

CrDem committed:
requested changes

CrDem committed:
diff cleaning

CrDem committed:
OptiXRender blas rebuild optimized

CrDem committed:
AS rebuild/refit optimized

CrDem committed:
merged latest Strelka improvementss

1. pulled and merged latest changes from upstream repo, replacing raw CUDA pointers to `std::unique_ptr`
2. there is no compation in TLAS update function now in order to improve performance

CrDem committed:
light meshes bug fixed

fixed crash triggered by scenes containing mesh-based light sources (meshes created specifically for lights in scene were omitted while creating optix meshes)

CrDem committed:
motion blur attempt

Initial motion blur implementation using matrix and vertex motion

Added support for:
- Matrix motion transforms in TLAS via `OptixMatrixMotionTransform`
- Vertex motion in BLAS with dual vertex buffers and `OptixMotionOptions`

Enabled in settings via `render/enableMotionBlur`.

CrDem committed:
partly fixed

CrDem committed:
deleted debug line

CrDem committed:
Motion blur works

- Fixed visual artifacts caused by incorrect normal interpolation during motion blur; vertex normals, tangents, and UVs are now properly interpolated between previous and current vertex states based on rayTime
- Added UI checkbox to enable or disable motion blur at runtime
- Complete motion blur disable via "render/enableMotionBlur" is now working properly
- Minor code cleanup

CrDem committed:
matrix motion bug fixed

Tackled an issue with matrix motion when only rotation channel was blurred.
Instances are now correctly stored at two time steps (previous and current) when motion blur is enabled.
Accumulation now resets on motion blur visibility toggle.

arhix52 commented on include/scene/scene.h:
```diff
@@ -92,17 +95,50 @@ class Scene
 
+    struct vertexSkinData //vertex skin data
+    {
+        glm::ivec4 joints{0};
+        glm::vec4 weights{0.0};
+        glm::float3 pos;
+        glm::float3 normal;
+        float pad0;
+        float pad1;
```
я бы тут поставил сначала pos, pad0, потом нормаль + пад1 чтобы доступы к ним (адреса) были тоже выровнены - влияет на перф

arhix52 commented on include/scene/scene.h:
```diff
@@ -92,17 +95,50 @@ class Scene
+
     struct Node
     {
+        enum class NodeType
```
`enum class NodeType : uint8_t {`

arhix52 commented on src/app/Editor.h:
```diff
@@ -141,26 +165,28 @@ class Editor
 
             m_display->onBeginFrame();
 
-            auto maxEDR = m_display->getMaxEDR();
```
это мерж конфликты или с чем-то связано?

arhix52 commented on src/app/Editor.h:
```diff
@@ -171,6 +197,31 @@ class Editor
         }
     }
 
+    void playAnimations(const double deltaTime) 
```
float?

arhix52 commented on src/app/Editor.h:
```diff
@@ -225,7 +276,9 @@ class Editor
                     camera.updateViewMatrix();
                     m_scene->addCamera(camera);
 
-                    m_sharedCtx = std::make_unique<SharedContext>();
```
merge conflict?

arhix52 commented on src/app/Editor.h:
```diff
@@ -410,19 +463,46 @@ class Editor
         m_settingsManager->setAs<uint32_t>("render/pt/tonemapperType", currentTonemapItemId);
 
-        auto gamma = m_settingsManager->getAs<float>("render/post/gamma");
+        float gamma = m_settingsManager->getAs<float>("render/post/gamma");
```
merge conflict?

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -235,115 +231,137 @@ OptiXRender::Curve* OptiXRender::createCurve(const oka::Curve& curve)
 
-    // Set up triangle input data
-    const CUdeviceptr vertexBuffer = mVertexBuffer->getPtr() + mesh.mVbOffset * sizeof(oka::Scene::Vertex);
+    CUdeviceptr vertexBuffer[2];
```
тут два буффера 0 для пред состояния - 1 для текущего состояния, можно это прописать в коментах и добавить константы чтобы использовать vertexBuffer[PREV_VB] = ... vertexBuffer[CURR_VB] = ...

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -191,10 +187,8 @@ OptiXRender::Curve* OptiXRender::createCurve(const oka::Curve& curve)
                                OPTIX_BUILD_FLAG_PREFER_FAST_TRACE;
     accel_options.operation = OPTIX_BUILD_OPERATION_BUILD;
 
-    const uint32_t pointsCount = mScene->getCurvesPoint().size(); // total points count in points buffer
```
merge conflicts

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -406,8 +475,43 @@ void OptiXRender::createTopLevelAccelerationStructure()
+            OptixTraversableHandle matrixMotionTransformHandle;
+
+            matrixMotionTransform.child = oi.traversableHandle;
+            matrixMotionTransform.motionOptions.numKeys   = 2;
```
много где встречается - можно в какую-то константу вынести

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -827,14 +1027,92 @@ void OptiXRender::updatePathtracerParams(const uint32_t width, const uint32_t he
+void OptiXRender::applySkinning()
+{
+    // joint matrices
+    {
```
лишние скобки?

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -827,14 +1027,92 @@ void OptiXRender::updatePathtracerParams(const uint32_t width, const uint32_t he
+        size_t jointMatSize = jointMat.size();
+        std::vector<sutil::Matrix4x4> cudaMatrices(jointMatSize);
+
+        std::transform(jointMat.begin(), jointMat.end(), cudaMatrices.begin(),
```
тут лучше просто for и заполнить матрицы?

arhix52 commented on src/render/optix/OptixRender.cpp:
```diff
@@ -827,14 +1027,92 @@ void OptiXRender::updatePathtracerParams(const uint32_t width, const uint32_t he
         createOptixMaterials();
         createPipeline();
         createVertexBuffer();
+        if (mEnableMotionBlur) createPrevBuffers();
```
скобки?

CrDem committed:
requested changes

arhix52 commented on src/app/Editor.h:
```diff
@@ -149,18 +175,21 @@ class Editor
             m_render->render(outputBuffer);
+            outputBuffer->map();
             oka::ImageBuffer outputImage;
-            outputImage.deviceData = outputBuffer->getDevicePointer();
+            outputImage.data = outputBuffer->getHostPointer();
+            outputImage.dataSize = outputBuffer->getHostDataSize();
             outputImage.height = outputBuffer->height();
             outputImage.width = outputBuffer->width();
             outputImage.pixel_format = oka::BufferFormat::FLOAT4;
-            outputImage.dataSize = outputBuffer->width() * outputBuffer->height() * outputBuffer->getElementSize();
             m_display->drawFrame(outputImage); // blit rendered image to swapchain
 
             drawUI(); // render ui to swapchain image in window resolution
             m_display->drawUI();
             m_display->onEndFrame(); // submit command buffer and present
 
+            outputBuffer->unmap();
+
```
похоже, что вот эти изменения относятся к лишним мемкопи

CrDem committed:
sync with upstream

Ported previously missed changes from the upstream repo. This resolves a number of local differences that had accumulated and were impacting performance.

CrDem committed:
motion blur button fixed

UI Fix: Hide "Motion Blur" button in "Animations" window when motion blur is completely disabled (render/enableMotionBlur is false).

arhix52 committed:
line endings + format
